File: frontend/src/lib/crypto/messages.ts
/**
 * Message crypto — matches the Rust messaging-crypto crate exactly.
 *
 * K_acte = HKDF(key=kMaster, info="notariat-msg-v1" || UUID(16), len=32)
 * K_send = HKDF(key=kActe, info="send" || SN(16), len=32)
 * encrypt: AES-256-GCM(K_send, plaintext, AAD=UUID(16)||ts(8 LE)||SN(16))
 * sign:    Ed25519(sk, SHA256(MSG_DOMAIN_TAG || ciphertext || nonce(12) || UUID(16) || ts(8 LE) || SN(16)))
 */

import { ed25519 } from '@noble/curves/ed25519.js';
import { hkdf } from '@noble/hashes/hkdf.js';
import { sha256 } from '@noble/hashes/sha2.js';
import { gcm } from '@noble/ciphers/aes.js';
import { randomBytes } from '@noble/ciphers/utils.js';
import { uuidToBytes, hexToBytes, timestampToLeBytes } from './keys';

const enc = new TextEncoder();

// Domain-separation tag for client message signatures — mirrors
// messaging_crypto::messages::MSG_DOMAIN_TAG in Rust (16 bytes incl. trailing NUL).
const MSG_DOMAIN_TAG = enc.encode('localpki-msg-v1\0');

// ─── Key derivation ───────────────────────────────────────────────────────────

export function deriveKActe(kMaster: Uint8Array, acteUuid: string): Uint8Array {
	const prefix = enc.encode('notariat-msg-v1');
	const uuidBytes = uuidToBytes(acteUuid);
	const info = new Uint8Array(prefix.length + 16);
	info.set(prefix, 0);
	info.set(uuidBytes, prefix.length);
	return hkdf(sha256, kMaster, undefined, info, 32);
}

export function deriveKSend(kActe: Uint8Array, snHex: string): Uint8Array {
	const prefix = enc.encode('send');
	const snBytes = hexToBytes(snHex);
	const info = new Uint8Array(prefix.length + 16);
	info.set(prefix, 0);
	info.set(snBytes, prefix.length);
	return hkdf(sha256, kActe, undefined, info, 32);
}

// ─── AAD and signing payload ──────────────────────────────────────────────────

function buildAad(acteUuid: string, timestamp: number, snHex: string): Uint8Array {
	const aad = new Uint8Array(40);
	aad.set(uuidToBytes(acteUuid), 0);
	aad.set(timestampToLeBytes(timestamp), 16);
	aad.set(hexToBytes(snHex), 24);
	return aad;
}

// Signing payload: MSG_DOMAIN_TAG || ciphertext || nonce(12) || UUID(16) || ts(8 LE) || SN(16)
// Mirrors messaging_crypto::messages::signing_payload in Rust.
function buildSigningPayload(
	ciphertext: Uint8Array,
	nonce: Uint8Array,
	acteUuid: string,
	timestamp: number,
	snHex: string
): Uint8Array {
	const uuidBytes = uuidToBytes(acteUuid);
	const tsBytes = timestampToLeBytes(timestamp);
	const snBytes = hexToBytes(snHex);
	const payload = new Uint8Array(MSG_DOMAIN_TAG.length + ciphertext.length + 12 + 16 + 8 + 16);
	let offset = 0;
	payload.set(MSG_DOMAIN_TAG, offset); offset += MSG_DOMAIN_TAG.length;
	payload.set(ciphertext, offset); offset += ciphertext.length;
	payload.set(nonce, offset);      offset += 12;
	payload.set(uuidBytes, offset);  offset += 16;
	payload.set(tsBytes, offset);    offset += 8;
	payload.set(snBytes, offset);
	return payload;
}

// ─── Encrypt / Decrypt ────────────────────────────────────────────────────────

export function encryptMessage(
	kActe: Uint8Array,
	plaintext: Uint8Array,
	acteUuid: string,
	senderSnHex: string,
	timestamp: number
): { ciphertext: Uint8Array; nonce: Uint8Array } {
	const kSend = deriveKSend(kActe, senderSnHex);
	const aad = buildAad(acteUuid, timestamp, senderSnHex);
	const nonce = randomBytes(12);
	const cipher = gcm(kSend, nonce, aad);
	const ciphertext = cipher.encrypt(plaintext);
	return { ciphertext, nonce };
}

export function decryptMessage(
	kActe: Uint8Array,
	senderSnHex: string,
	ciphertext: Uint8Array,
	nonce: Uint8Array,
	acteUuid: string,
	timestamp: number
): Uint8Array {
	const kSend = deriveKSend(kActe, senderSnHex);
	const aad = buildAad(acteUuid, timestamp, senderSnHex);
	const cipher = gcm(kSend, nonce, aad);
	return cipher.decrypt(ciphertext);
}

// ─── Sign / Verify ────────────────────────────────────────────────────────────

export function signMessage(
	signingKey: Uint8Array,
	ciphertext: Uint8Array,
	nonce: Uint8Array,
	acteUuid: string,
	senderSnHex: string,
	timestamp: number
): Uint8Array {
	const payload = buildSigningPayload(ciphertext, nonce, acteUuid, timestamp, senderSnHex);
	const digest = sha256(payload);
	return ed25519.sign(digest, signingKey);
}

export function verifyMessageSignature(
	verifyingKey: Uint8Array,
	ciphertext: Uint8Array,
	nonce: Uint8Array,
	acteUuid: string,
	senderSnHex: string,
	timestamp: number,
	signature: Uint8Array
): boolean {
	try {
		const payload = buildSigningPayload(ciphertext, nonce, acteUuid, timestamp, senderSnHex);
		const digest = sha256(payload);
		return ed25519.verify(signature, digest, verifyingKey);
	} catch {
		return false;
	}
}

File: crates/messaging-crypto/src/messages.rs
// Message encryption and signing.
//
// Send (Alice): derive K_send → AES-256-GCM(K_send, M, AAD) + Ed25519.Sign(sk, H(C || context))
// Receive (Bob): verify signature on ciphertext → derive K_send_Alice → decrypt
//
// The signature is computed over the *ciphertext*, not the plaintext. This lets the server
// reject forgeries without ever reading the message. Non-repudiation on the cleartext is
// preserved because AES-GCM is AEAD: a given (ciphertext, nonce, AAD, key) tuple decrypts
// to exactly one plaintext or fails. See ARCHITECTURE.md §5.3.

use aes_gcm::{aead::{Aead, KeyInit, Payload}, Aes256Gcm};
use ed25519_dalek::{ed25519::signature::Signer, Verifier};
use rand::{RngCore, rngs::OsRng};
use sha2::{Digest, Sha256};

use crate::{error::CryptoError, keys::derive_k_send};
use localpki_core::cert::SerialNumber;
use serde::{Deserialize, Serialize};

/// Encrypted message stored server-side. The server never sees plaintext.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EncryptedMessage {
    pub ciphertext: Vec<u8>,
    pub nonce: [u8; 12],
    /// Ed25519(sk_sender, SHA256(MSG_DOMAIN_TAG || ciphertext || nonce || acte_uuid || timestamp || SN_sender))
    pub signature: ed25519_dalek::Signature,
    /// Monotonic sequence number within the acte, assigned by the server.
    pub seq: u64,
}

pub fn encrypt_message(
    k_acte: &[u8; 32],
    plaintext: &[u8],
    acte_uuid: &uuid::Uuid,
    sender_sn: &SerialNumber,
    timestamp: i64,
) -> Result<(Vec<u8>, [u8; 12]), CryptoError> {
    let k_send = derive_k_send(k_acte, sender_sn);
    let aad = build_aad(acte_uuid, timestamp, sender_sn);
    let mut nonce_bytes = [0u8; 12];
    OsRng.fill_bytes(&mut nonce_bytes);

    let ciphertext = Aes256Gcm::new_from_slice(&k_send)
        .map_err(|_| CryptoError::KeyDerivation)?
        .encrypt(
            aes_gcm::Nonce::from_slice(&nonce_bytes),
            Payload { msg: plaintext, aad: &aad },
        )
        .map_err(|_| CryptoError::Encryption)?;

    Ok((ciphertext, nonce_bytes))
}

pub fn decrypt_message(
    k_acte: &[u8; 32],
    sender_sn: &SerialNumber,
    ciphertext: &[u8],
    nonce: &[u8; 12],
    acte_uuid: &uuid::Uuid,
    timestamp: i64,
) -> Result<Vec<u8>, CryptoError> {
    let k_send = derive_k_send(k_acte, sender_sn);
    let aad = build_aad(acte_uuid, timestamp, sender_sn);

    Aes256Gcm::new_from_slice(&k_send)
        .map_err(|_| CryptoError::KeyDerivation)?
        .decrypt(
            aes_gcm::Nonce::from_slice(nonce),
            Payload { msg: ciphertext, aad: &aad },
        )
        .map_err(|_| CryptoError::Decryption)
}

pub fn sign_message(
    signing_key: &ed25519_dalek::SigningKey,
    ciphertext: &[u8],
    nonce: &[u8; 12],
    acte_uuid: &uuid::Uuid,
    sender_sn: &SerialNumber,
    timestamp: i64,
) -> ed25519_dalek::Signature {
    signing_key.sign(&Sha256::digest(&signing_payload(ciphertext, nonce, acte_uuid, timestamp, sender_sn)))
}

pub fn verify_message_signature(
    verifying_key: &ed25519_dalek::VerifyingKey,
    ciphertext: &[u8],
    nonce: &[u8; 12],
    acte_uuid: &uuid::Uuid,
    sender_sn: &SerialNumber,
    timestamp: i64,
    signature: &ed25519_dalek::Signature,
) -> Result<(), CryptoError> {
    verifying_key
        .verify(&Sha256::digest(&signing_payload(ciphertext, nonce, acte_uuid, timestamp, sender_sn)), signature)
        .map_err(|_| CryptoError::InvalidMessageSignature)
}

/// AAD (Additional Authenticated Data): acte_uuid (16) || timestamp (8, LE) || SN (16) — binds ciphertext to its context.
fn build_aad(acte_uuid: &uuid::Uuid, timestamp: i64, sn: &SerialNumber) -> Vec<u8> {
    let mut aad = Vec::with_capacity(40);
    aad.extend_from_slice(acte_uuid.as_bytes());
    aad.extend_from_slice(&timestamp.to_le_bytes());
    aad.extend_from_slice(&sn.0);
    aad
}

/// Domain-separation tag for client message signatures. Keeps a message
/// signature from being reusable in another user-key context (SI over the cert
/// DER, participant-add, revocation). Mirror in frontend/src/lib/crypto/messages.ts.
pub const MSG_DOMAIN_TAG: &[u8] = b"localpki-msg-v1\0";

/// Signing payload: tag || ciphertext || nonce (12) || acte_uuid (16) || timestamp (8, LE) || SN (16).
fn signing_payload(
    ciphertext: &[u8],
    nonce: &[u8; 12],
    acte_uuid: &uuid::Uuid,
    timestamp: i64,
    sn: &SerialNumber,
) -> Vec<u8> {
    let mut payload = Vec::with_capacity(MSG_DOMAIN_TAG.len() + ciphertext.len() + 52);
    payload.extend_from_slice(MSG_DOMAIN_TAG);
    payload.extend_from_slice(ciphertext);
    payload.extend_from_slice(nonce);
    payload.extend_from_slice(acte_uuid.as_bytes());
    payload.extend_from_slice(&timestamp.to_le_bytes());
    payload.extend_from_slice(&sn.0);
    payload
}
